File: fastapi/routes/checkout.py
import uuid
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List

from core.dependencies import get_db
from models.checkout import Checkout, CheckoutCreate, CheckoutItem
from services import checkout_service
from core.dependencies import get_current_user
from models.user import User as UserModel

router = APIRouter(prefix="/checkouts", tags=["Checkouts"])

def _checkout_item_to_dict(item: CheckoutItem) -> dict:
    return {
        "itemId": item.item_id,
        "bookId": item.book_id,
        "ownerId": item.owner_id,
        "actionType": item.action_type,
        "price": float(item.price) if item.price is not None else None,
        "deposit": float(item.deposit) if item.deposit is not None else None,
        "shippingMethod": item.shipping_method,
        "shippingQuote": float(item.shipping_quote) if item.shipping_quote is not None else None,
    }


def _checkout_to_dict(checkout: Checkout) -> dict:
    return {
        "checkoutId": checkout.checkout_id,
        "userId": checkout.user_id,
        "contactName": checkout.contact_name,
        "phone": checkout.phone,
        "street": checkout.street,
        "city": checkout.city,
        "postcode": checkout.postcode,
        "state": checkout.state,
        "country": checkout.country,
        "deposit": float(checkout.deposit),
        "serviceFee": float(checkout.service_fee),
        "bookFee": float(checkout.book_fee),
        "shippingFee": float(checkout.shipping_fee),
        "totalDue": float(checkout.total_due),
        "status": checkout.status,
        "createdAt": checkout.created_at.isoformat(),
        "updatedAt": checkout.updated_at.isoformat(),
        "items": [_checkout_item_to_dict(item) for item in checkout.items],
    }


@router.put("/{checkout_id}", status_code=status.HTTP_200_OK)
async def update_checkout(
    checkout_id: str,
    checkout_in: CheckoutCreate,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_user),
):
    """Update an existing checkout if it's still PENDING"""
    updated = await checkout_service.update_checkout(db, checkout_id, checkout_in)
    if not updated:
        raise HTTPException(status_code=404, detail="Checkout not found")
    return _checkout_to_dict(updated)

@router.get("/list", status_code=status.HTTP_200_OK)
def list_my_checkouts(
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_user),
):
    """List all checkouts for the current logged-in user"""
    checkouts = checkout_service.get_pending_checkouts_by_user_id(
        db, current_user.user_id)
    if not checkouts:
        raise HTTPException(
            status_code=404, detail="No checkouts found for this user")
    return [_checkout_to_dict(c) for c in checkouts]


@router.post("/create", status_code=status.HTTP_201_CREATED)
async def create_checkout(
    checkout_in: CheckoutCreate,
    db: Session = Depends(get_db),
    current_user: UserModel = Depends(get_current_user),
):
    """Create a new checkout with items"""
    checkout_in.userId = current_user.user_id
    checkout = await checkout_service.create_checkout(db, checkout_in)
    return _checkout_to_dict(checkout)


@router.delete("/{checkout_id}", status_code=status.HTTP_200_OK)
def delete_checkout(checkout_id: str, db: Session = Depends(get_db)):
    """Delete a checkout"""
    deleted = checkout_service.delete_checkout(db, checkout_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Checkout not found")
    return {"deletedId": checkout_id}

File: frontendNext/utils/checkout.ts
import axios from "axios";
import { getToken, getApiUrl } from "./auth";
import type { User } from "@/app/types/user";


const API_URL = getApiUrl();

// -------- check checkout list --------
export async function getMyCheckouts() {
  const res = await fetch(`${API_URL}/api/v1/checkouts/list`, {
    method: "GET",
    headers: {
      Authorization: `Bearer ${localStorage.getItem("access_token")}`,
    },
    credentials: "include",
  });

  if (!res.ok) {
    if (res.status === 404) return []; // 没有 checkout
    throw new Error("Failed to fetch checkouts");
  }
  
  return res.json();
}

// -------- delete checkout --------
export async function deleteCheckout(checkoutId: string) {
  const res = await fetch(`${API_URL}/api/v1/checkouts/${checkoutId}`, {
    method: "DELETE",
    headers: {
      Authorization: `Bearer ${localStorage.getItem("access_token")}`,
    },
    credentials: "include",
  });

  if (!res.ok) {
    throw new Error("Failed to delete checkout");
  }
  return res.json();
}

// -------- create checkout --------
export async function createCheckout(payload: any) {
  try {
    const res = await axios.post(`${API_URL}/api/v1/checkouts/create`, payload, {
      headers: {
        Authorization: `Bearer ${localStorage.getItem("access_token")}`,
      },
      withCredentials: true,
    });
    return res.data;
  } catch (err: any) {
    if (err.response) {
      console.error("Create checkout failed:", err.response.data);
      throw new Error("Failed to create checkout: " + JSON.stringify(err.response.data));
    }
    throw new Error("Failed to create checkout");
  }
}

// -------- Rebuild checkout ( update ) --------
export async function rebuildCheckout(
  user: User,
  items: any[],
  shipping: Record<string, "post" | "pickup"> = {},
  selectedQuotes: Record<string, any> = {}
) {
  // 1. get current checkout id
  const existing = await getMyCheckouts();
  const checkoutId = existing.length > 0 ? existing[0].checkoutId : null;

  // 2. build payload
  const payload = {
    userId: user.id,
    contactName: user.name || "",
    phone: user.phoneNumber || "",
    street: user.streetAddress || "",
    city: user.city || "",
    state: user.state || "",
    postcode: user.zipCode || "",
    country: "Australia",

    items: items.map((it) => {
      const bookId = it.bookId || it.id;   // ✅ 兼容 cart 和 checkout

      if (!bookId) {
        console.error("Missing bookId in item:", it);
        throw new Error("Item missing bookId, cannot create checkout");
      }
      const mode = it.mode || it.actionType?.toLowerCase();
      const quote = selectedQuotes[it.ownerId];

      console.log("Owner:", it.ownerId, "mode:", mode, "Quote:", quote);

      return {
        itemId: it.itemId || it.cartItemId,
        bookId,
        ownerId: it.ownerId,
        actionType: (it.mode || it.actionType)?.toUpperCase(), // default BORROW
        price: mode === "purchase" ? (it.salePrice || it.price) : undefined, // Sale Price
        deposit: mode === "borrow" ? it.deposit : undefined,
        shippingMethod:
          shipping[bookId] ||
          it.shippingMethod ||
          "",
        serviceCode:
          quote?.serviceLevel === "Express"
            ? "AUS_PARCEL_EXPRESS"
            : "AUS_PARCEL_REGULAR",
        shippingQuote: quote?.cost || 0,
      };
    }),
  };

  // 3. if checkout exist and pending → update
  if (checkoutId) {
    try {
      const res = await axios.put(
        `${API_URL}/api/v1/checkouts/${checkoutId}`,
        payload,
        { headers: { Authorization: `Bearer ${localStorage.getItem("access_token")}` } }
      );
      console.log("Payload sent to updateCheckout:", payload);

      return res.data;
    } catch (err: any) {
      if (err.response?.status === 404) {
        // checkout doesn't exist then create
        return await createCheckout(payload);
      }
      throw err;
    }
  }

  // 4. if no checkout → create
  console.log("Payload sent to createCheckout:", payload);
  const newCheckout = await createCheckout(payload);
  console.log("Checkout returned from backend:", newCheckout);
  return newCheckout;
}
